### PR TITLE
ctest should not leak console handle, should attempt to set the color…

### PR DESCRIPTION
**Pull request by anporumb**
…ing mode only when getting it succeeded before and should revert the coloring mode the initial value

### CODE REVIEW

File: src/ctest.c [hunk 1]
```diff
@@ -201,6 +214,25 @@ size_t RunTests(const TEST_FUNCTION_DATA* testListHead, const char* testSuiteNam
...
+            if (!SetConsoleMode(std_out_handle, console_mode_initial))
+            {
+                (void)printf("Error resetting console mode to initial value of %" PRIx32 ". GetLastError()=%" PRIx32 "\n", console_mode_initial, GetLastError());
+            }
+        }
+
+        if (!CloseHandle(std_out_handle))
+        {
+            (void)printf("failed to CloseHandle %p. GetLastError()=%" PRId32 "\n", std_out_handle, GetLastError());
+        }
```
Comment (dcristoloveanu) [inline src/ctest.c]: handles returned with GetStdHandle are special.
I actually did not CloseHandle on purpose based on mny reading.
They are global handles (same handle being returned at any call of GetStdHandle). Our code here is not owning that handle so it should not close it.

All MSDN examples showing usage of APIs that act on handles returned by GetStdHandle do not have CloseHandle in the examples code either (for example: [https://docs.microsoft.com/en-us/windows/console/using-the-high-level-input-and-output-functions](https://docs.microsoft.com/en-us/windows/console/using-the-high-level-input-and-output-functions)).

Here is the MSDN page that I also read which makes me think they should not have CloseHandle called on them:
[https://docs.microsoft.com/en-us/windows/console/console-handles](https://docs.microsoft.com/en-us/windows/console/console-handles)

I also did this simple console app experiment which seems to confirm my understanding:

Code:

#include <stdio.h>
#include "windows.h"

int main(void)
{
    HANDLE out_handle = GetStdHandle(STD_OUTPUT_HANDLE);
    BOOL close_result = printf("handle 1st time is %p\r\n", out_handle);
    if (!CloseHandle(out_handle))
    {
        fprintf(stderr, "Error closing handle 1st time, lasterror = %d\r\n", GetLastError());
    }
    HANDLE out_handle_2 = GetStdHandle(STD_OUTPUT_HANDLE);
    printf("handle 2nd time is %p\r\n", out_handle_2);
    if (!CloseHandle(out_handle_2))
    {
        fprintf(stderr, "Error closing handle 2nd time, lasterror = %d\r\n", GetLastError());
    }
}

Result:

handle 1st time is 00000054
Error closing handle 2nd time, lasterror = 6
Press any key to continue . . .

File: src/ctest.c [hunk 2]
```diff
@@ -201,6 +214,25 @@ size_t RunTests(const TEST_FUNCTION_DATA* testListHead, const char* testSuiteNam
...
+#if defined _MSC_VER
+    if (std_out_handle != INVALID_HANDLE_VALUE)
+    {
+        if (SetConsoleMode_succeeded)
+        {
+            /*revert console to initial state*/
+            if (!SetConsoleMode(std_out_handle, console_mode_initial))
+            {
+                (void)printf("Error resetting console mode to initial value of %" PRIx32 ". GetLastError()=%" PRIx32 "\n", console_mode_initial, GetLastError());
+            }
```
Comment (dcristoloveanu) [inline src/ctest.c]: This is nice, should have done that in the initial checkin

File: src/ctest.c [hunk 3]
```diff
@@ -26,22 +30,31 @@ size_t RunTests(const TEST_FUNCTION_DATA* testListHead, const char* testSuiteNam
...
-    {
-        (void)printf("Error getting console mode, no coloring available");
-    }
-
-    console_mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
-    if (!SetConsoleMode(std_out_handle, console_mode))
+    else
     {
-        (void)printf("Error setting console mode, no coloring available");
+        if (!GetConsoleMode(std_out_handle, &console_mode_initial))
```
Comment (dcristoloveanu) [inline src/ctest.c]: >GetConsoleMode [](start = 13, length = 14)

Cheers for fixing this to only execute on previous call's success.

File: src/ctest.c (hunk 1 above)
Comment (anporumb) [reply src/ctest.c]: Alright, alright :)

---
In reply to: [291295480](https://github.com/Azure/azure-ctest/pull/29#discussion_r291295480) [](ancestors = 291295480)

**Pull request merged**